### PR TITLE
✨ Move PROS 4 to early access

### LINKED ISSUE #295: SSL issues
#### Expected Behavior:
connect and updated

#### Actual Behavior:
it errored

#### Steps to reproduce:
Not sure i just installed pros

#### System information:
Platform: V5
Name      Version    Origin
--------  ---------  -------------
kernel    3.8.0      pros-mainline
okapilib  4.8.0      pros-mainline

#### Screenshots/Output Dumps/Stack Traces
```diff
PS C:\Users\1300286\over-under> pros conduct upgrade
Upgrading kernel
Updating pros-mainline... ERROR - __main__:main - HTTPSConnectionPool(host='purduesigbots.github.io', port=443): Max retries exceeded with url: /pros-mainline/pros-mainline.json (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)'))) - pros-cli version:3.4.3
  File "requests\sessions.py", line 703, in send
  File "requests\adapters.py", line 517, in send
requests.exceptions.SSLError: HTTPSConnectionPool(host='purduesigbots.github.io', port=443): Max retries exceeded with url: /pros-mainline/pros-mainline.json (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)')))
We detected something went wrong! Do you want to send a report? [y/N]: y
Sending bug report.
Want to get updates? Visit https://pros.cs.purdue.edu/report.html?event=980c477312d94b0388cd27e7ac0195ae
WARNING - urllib3.connectionpool:urlopen - Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)'))': /api/1226033/store/ - pros-cli version:3.4.3
Sentry is attempting to send 2 pending events
Waiting up to 2 seconds
Press Ctrl-Break to quit
WARNING - urllib3.connectionpool:urlopen - Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)'))': /api/1226033/store/ - pros-cli version:3.4.3
WARNING - urllib3.connectionpool:urlopen - Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)'))': /api/1226033/store/ - pros-cli version:3.4.3
WARNING - urllib3.connectionpool:urlopen - Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)'))': /api/1226033/envelope/ - pros-cli version:3.4.3
WARNING - urllib3.connectionpool:urlopen - Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)'))': /api/1226033/envelope/ - pros-cli version:3.4.3
WARNING - urllib3.connectionpool:urlopen - Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)'))': /api/1226033/envelope/ - pros-cli version:3.4.3
```

### PR DESCRIPTION
**Pull request by 12944qwerty**
#### Summary:
Deprecate PROS v3.

 - Renames beta to modern or v4
 - warns user v3 is deprecated once

Closes #295

### CODE REVIEW

File: pros/cli/conductor.py [hunk 1]
```diff
@@ -90,8 +90,8 @@ def fetch(query: c.BaseTemplate):
               help="Force apply the template, disregarding if the template is already installed.")
 @click.option('--remove-empty-dirs/--no-remove-empty-dirs', 'remove_empty_directories', is_flag=True, default=True,
               help='Remove empty directories when removing files')
-@click.option('--beta', is_flag=True, default=False, show_default=True,
-              help='Allow applying beta templates')
+@click.option('--modern', is_flag=True, default=False, show_default=True,
```
Comment (ayushuk) [inline pros/cli/conductor.py]: There's a few functions here that have the beta flag but dont have a parameter or usage of them. Can you check if its actually being used? If not just remove the flag.

File: pros/conductor/conductor.py [hunk 2]
```diff
@@ -17,11 +18,11 @@
 
 MAINLINE_NAME = 'pros-mainline'
 MAINLINE_URL = 'https://purduesigbots.github.io/pros-mainline/pros-mainline.json'
-BETA_NAME = 'kernel-beta-mainline'
-BETA_URL = 'https://raw.githubusercontent.com/purduesigbots/pros-mainline/master/beta/kernel-beta-mainline.json'
+PROS4_NAME = 'kernel-4-mainline'
```
Comment (ayushuk) [inline pros/conductor/conductor.py]: Lets use a different naming scheme. This represents a depo and as of right now we call that depo the beta depo. I dont think `PROS4` would be the best way to reference the depo in code. I'm thinking about naming the current depo to something like `early access` so that references would look like `early_access_templates` or `depo[EARLY_ACCESS]`. Seems a lot more professional and will avoid future confusion.

File: pros/conductor/conductor.py [hunk 3]
```diff
@@ -230,18 +232,25 @@ def apply_template(self, project: Project, identifier: Union[str, BaseTemplate],
...
                         confirm = ui.confirm(f'Warning! Upgrading project to PROS 4 will cause breaking changes. '
-                                             f'For PROS 4 LLEMU/LVGL to function, the library liblvgl is required. '
-                                             f'Run \'pros conductor apply liblvgl --beta\' in the project directory. '
                                              f'Do you still want to upgrade?')
                         if not confirm:
                             raise dont_send(
                                 InvalidTemplateException(f'Not upgrading'))
                     if template.version[0] == '3' and curr_proj.kernel[0] == '4':
-                        confirm = ui.confirm(f'Warning! Downgrading project to PROS 3 will cause breaking changes. '
+                        confirm = ui.confirm(f'PROS 3 is now deprecated. Downgrading project to PROS 3 will also cause breaking changes. '
```
Comment (ayushuk) [inline pros/conductor/conductor.py]: PROS 3 is not deprecated, its going to be the default for almost another year.

File: pros/conductor/conductor.py [hunk 4]
```diff
@@ -230,18 +232,25 @@ def apply_template(self, project: Project, identifier: Union[str, BaseTemplate],
...
                                              f'Do you still want to downgrade?')
                         if not confirm:
                             raise dont_send(
                                 InvalidTemplateException(f'Not downgrading'))
+            elif template.version[0] == '3' and not self.warned_pros3_deprecated:
+                confirm = ui.confirm(f'PROS 3 is now deprecated. PROS 4 is now the latest version. '
```
Comment (ayushuk) [inline pros/conductor/conductor.py]: same as above, lets be careful with wording

File: pros/conductor/conductor.py [hunk 5]
```diff
@@ -279,13 +288,15 @@ def remove_template(project: Project, identifier: Union[str, BaseTemplate], remo
                                     remove_empty_directories=remove_empty_directories)
 
     def new_project(self, path: str, no_default_libs: bool = False, **kwargs) -> Project:
-        self.is_beta = kwargs.get('beta', False)
+        self.use_pros4 = kwargs.get('modern', False)
+        if not self.use_pros4 and self.warned_pros3_deprecated:
+            ui.echo(f"PROS 3 is deprecated. PROS 4 is now the latest version. "
```
Comment (ayushuk) [inline pros/conductor/conductor.py]: maybe here we can notify that PROS 4 is available as an option but PROS 3 is still the default

File: pros/conductor/conductor.py [hunk 6]
```diff
@@ -279,13 +288,15 @@ def remove_template(project: Project, identifier: Union[str, BaseTemplate], remo
...
+                    f"To use PROS 4 use the --modern flag.")
+
         if Path(path).exists() and Path(path).samefile(os.path.expanduser('~')):
             raise dont_send(ValueError('Will not create a project in user home directory'))
-        for char in str(Path(path)):
-            if char in ['?', '<', '>', '*', '|', '^', '#', '%', '&', '$', '+', '!', '`', '\'', '=',
-                        '@', '\'', '{', '}', '[', ']', '(', ')', '~'] or ord(char) > 127:
-                raise dont_send(ValueError(f'Invalid character found in directory name: \'{char}\''))
+        if re.match(r'^[\w\-. /]+$', str(Path(path))) is None:
+            raise dont_send(ValueError('Invalid characters found in path'))
```
Comment (ayushuk) [inline pros/conductor/conductor.py]: nice improvement, lets put this in another PR we can track the changes in the next update

File: pros/conductor/conductor.py [hunk 7]
```diff
@@ -300,23 +311,13 @@ def new_project(self, path: str, no_default_libs: bool = False, **kwargs) -> Pro
...
-            else:
-                for library in self.default_libraries[proj.target]:
-                    try:
-                        # remove kernel version so that latest template satisfying query is correctly selected
-                        if 'version' in kwargs:
-                            kwargs.pop('version')
-                        self.apply_template(proj, library, **kwargs)
-                    except Exception as e:
-                        logger(__name__).exception(e)
+            libraries = self.pros4_libraries if self.use_pros4 else self.default_libraries
```
Comment (ayushuk) [inline pros/conductor/conductor.py]: nice job! I like to see code cleanup as we go

File: pros/cli/conductor.py (hunk 1 above)
Comment (ayushuk) [reply pros/cli/conductor.py]: For mentions check if its accessed through kwargs ^^

**Review by ayushuk — CHANGES_REQUESTED** [review]
For CLI changes, we usually required you to test your changes with a test plan expected and actual outputs on the PR. Overall it looks good, good job.

**Comment by ayushuk** [comment]
Closes #295

**Comment by ayushuk** [comment]
Create new pros 3 project without early access flag for the first time:
new pros 3 project first time

Create a new pros 3 project after the first time without the flag:
new pros 3 project second time

Create a new pros 4 project with the early access flag:
new pros project early access flag

Create a new pros project after early access is enabled:
pros 4 project no flag

File: pros/cli/conductor.py [hunk 8]
```diff
@@ -206,8 +200,8 @@ def uninstall_template(project: c.Project, query: c.BaseTemplate, remove_user: b
               help='Compile the project after creation')
 @click.option('--build-cache', is_flag=True, default=None, show_default=False,
               help='Build compile commands cache after creation. Overrides --compile-after if both are specified.')
-@click.option('--beta', is_flag=True, default=False, show_default=True,
-              help='Create a project with a beta template')
+@click.option('--early-access', '--early', '-ea', 'early_access', '--beta', is_flag=True, default=False, show_default=True,
+              help='Create a project with a PROS v4 template')
```
Comment (BennyBot) [inline pros/cli/conductor.py]:
```suggestion
              help='Create a project using the PROS 4 kernel')
```

**Review by ayushuk — APPROVED** [review]

File: pros/cli/conductor.py [hunk 9]
```diff
@@ -256,13 +250,13 @@ def new_project(ctx: click.Context, path: str, target: str, version: str,
               help='Force update all remote depots, ignoring automatic update checks')
 @click.option('--limit', type=int, default=15,
               help='The maximum number of displayed results for each library')
-@click.option('--beta', is_flag=True, default=False, show_default=True,
-              help='View beta templates in the listing')
+@click.option('--early-access', '--early', '-ea', '--beta', 'early_access', is_flag=True, default=False, show_default=True,
+              help='View PROS v4 templates in the listing')
```
Comment (BennyBot) [inline pros/cli/conductor.py]:
```suggestion
              help='View a list of early access templates')
```

**Review by BennyBot — DISMISSED** [review]

**Comment by noam987** [comment]
Only concern I have is the okapi Message (also We should move okapi out of being a default template)

**Review by ayushuk — CHANGES_REQUESTED** [review]
Need to fix the okapi error

**Comment by ayushuk** [comment]
There are some issues I have with the current flow. 

1. Saying "no" to the initial prompt that says "PROS 3 is available" just raises an error and ends the execution. When you try to create a new project the directory is already created so then you try to make a project again and it gives you an error that a project already exists even though its just a directory with `project.pros` so then you have the delete that directory to make a project. When you make a new project after deleting the directory it makes a project in PROS 3 after the user already said no to using PROS 3. Seeing an error after saying no is probably just going to cause confusion and I think when someone says no to the prompt a PROS 4 project should just be created.
2. There isn't any way (other than deleting `conductor.pros`) to turn off early access after you already enabled it. For someone that accidentally enabled early access we should have a way for them to turn off early access. This can probably be a flag. We dont want people going through hoops to reverse an action they accidentally did.
3. I tried to make a PROS 3 project with early access enabled and I got an error where it tried to apply LVGL. The LVGL template shouldn't be trying to be applied to PROS 3. We should fix this to avoid confusion.


Screenshot 2023-11-13 at 8 59 28 AM
Screenshot 2023-11-13 at 9 02 25 AM

File: pros/cli/conductor.py [hunk 10]
```diff
@@ -91,8 +91,8 @@ def fetch(query: c.BaseTemplate):
               help="Force apply the template, disregarding if the template is already installed.")
 @click.option('--remove-empty-dirs/--no-remove-empty-dirs', 'remove_empty_directories', is_flag=True, default=True,
               help='Remove empty directories when removing files')
-@click.option('--beta', is_flag=True, default=False, show_default=True,
-              help='Allow applying beta templates')
+@click.option('--early-access/--no-early-access', '--early/--no-early', '-ea/-nea', 'early_access', '--beta/--no-beta', default=None,
```
Comment (ayushuk) [inline pros/cli/conductor.py]: imo `--disable-early-access` > `--no-early-access` but open to more opinions

**Review by ayushuk — COMMENTED** [review]
Code lgtm, will test this out tomorrow I think we should be good to merge

**Comment by ayushuk** [comment]
Some things after testing this.
1. The disable early access feature works very well for me, tried some edge cases and it performs as I expect
2. The problem I mentioned in point 1 of the large comment above about raising an error and creating a directory is still partially an issue. When you choose "no", an empty project directory is created, while you can make a new project with the same name, its still an issue to be because the directory still exists. For this reason, I think when someone selects the "no" option to "do you want to use pros 3" then you should just go ahead and make a PROS 4 project instead of having to run another command since we already know their intention.
3. Point 3 in the large comment is still an issue

**Review by ayushuk — APPROVED** [review]
LGTM
